airbyte-cdk/python/airbyte_cdk/sources/declarative/async_job/job_orchestrator.py

```diff
@@ -198,13 +231,43 @@ def _process_running_partitions_and_yield_completed_ones(self) -> Generator[Asyn
                 case AsyncJobStatus.RUNNING:
                     current_running_partitions.append(partition)
                 case _ if partition.has_reached_max_attempt():
+                    self._stop_partition(partition)
                     self._process_partitions_with_errors(partition)
                 case _:
+                    self._stop_timed_out_jobs(partition)
+
                     # job will be restarted in `_start_job`
                     current_running_partitions.insert(0, partition)
+
+            for job in partition.jobs:
+                # We only remove completed jobs as we want failed/timed out jobs to be re-allocated in priority
+                if job.status() == AsyncJobStatus.COMPLETED:
+                    self._job_tracker.remove_job(job.api_job_id())
+
         # update the referenced list with running partitions
         self._running_partitions = current_running_partitions
 
+    def _stop_partition(self, partition: AsyncPartition) -> None:
+        for job in partition.jobs:
+            if job.status() in {AsyncJobStatus.RUNNING, AsyncJobStatus.TIMED_OUT}:
```

Comment:
nit: I think this is creating a new set on every partition jobs

Reply:
I'll extract that